src/components/timeline/LayerDiscrete.svelte

```diff
@@ -100,59 +110,82 @@
 
   let canvas: HTMLCanvasElement;
   let ctx: CanvasRenderingContext2D;
+  let colorCache: Record<string, string> = {};
   let dragCurrentX: number | null = null;
   let dragOffsetX: number | null = null;
   let dragPreviousX: number | null = null;
   let dragActivityDirectiveActive: ActivityDirective | null = null;
   let dragStartX: number | null = null;
   let maxActivityWidth: number;
+  let maxExternalEventWidth: number;
   let minRectSize: number = 4;
   let planStartTimeMs: number;
   let quadtreeActivityDirectives: Quadtree<QuadtreeRect>;
   let quadtreeSpans: Quadtree<QuadtreeRect>;
+  let quadtreeExternalEvents: Quadtree<QuadtreeRect>;
   let visibleActivityDirectivesById: Record<ActivityDirectiveId, ActivityDirective> = {};
   let visibleSpansById: Record<SpanId, Span> = {};
-  let colorCache: Record<string, string> = {};
+  let visibleExternalEventsById: Record<ExternalEventId, ExternalEvent> = {};
 
   // Asset cache
   const assets: { anchorIcon: HTMLImageElement | null } = { anchorIcon: null };
   const textMetricsCache: Record<string, TextMetrics> = {};
 
   $: onBlur(blur);
   $: onContextmenu(contextmenu);
-  $: onDblclick(dblclick);
+  $: onDblClick(dblclick);
   $: onFocus(focus);
   $: onMousedown(mousedown);
   $: onMousemove(mousemove);
   $: onMouseout(mouseout);
-  $: onMouseup(mouseup);
+  $: onMouseUp(mouseup);
 
   $: anchorIconWidth = 16;
   $: anchorIconMarginLeft = 4;
   $: canvasHeightDpr = drawHeight * dpr;
   $: canvasWidthDpr = drawWidth * dpr;
-  $: rowHeight = activityOptions.activityHeight + (activityOptions.displayMode === 'compact' ? 0 : 0);
+  $: rowHeight = discreteOptions.height + (discreteOptions.displayMode === 'compact' ? 0 : 0);
   $: timelineLocked = timelineLockStatus === TimelineLockStatus.Locked;
   $: planStartTimeMs = getUnixEpochTime(getDoyTime(new Date(planStartTimeYmd)));
 
-  $: if (
+  // the following are NOT mutually exclusive.
+  $: canDrawActivities =
     activityDirectives &&
     showDirectives !== undefined &&
     showSpans !== undefined &&
+    selectedActivityDirectiveId !== undefined &&
+    selectedSpanId !== undefined &&
+    spansMap &&
+    discreteOptions.activityOptions &&
+    spans;
+
+  $: canDrawExternalEvents =
+    externalEvents && selectedExternalEventId !== undefined && discreteOptions.externalEventOptions;
+
+  $: if (
     canvasHeightDpr &&
     canvasWidthDpr &&
     ctx &&
     drawHeight &&
     drawWidth &&
     dpr &&
-    selectedActivityDirectiveId !== undefined &&
-    selectedSpanId !== undefined &&
-    spansMap &&
-    activityOptions &&
+    discreteOptions &&
     viewTimeRange &&
     xScaleView &&
-    spans &&
-    activityTree
+    discreteTree &&
+    (canDrawActivities || canDrawExternalEvents)
+  ) {
+    draw();
+  }
+
+  // force a redraw as a reaction to a new selection, else a new selection won't update anything. TODO: make this more efficient! Redraw specific items, by matching ids?
```

Comment:
Reason for the change from the old behavior of checking that selected ids weren't undefined? That should have worked for reacting to selection. Doesn't address the TODO but that'd be difficult without an additional highlight canvas layer that could be drawn separately from the main canvas layer here.

Reply:
My main issue with this approach here is that the timeline will draw even if there are no directives, events, canvas height, time range, etc.

Reply:
I believe the reasoning here is that after everything is deselected, we would want to trigger a redraw. That being said, no discernible behavior change was found on refactoring it per your suggestion (see commit "[update selection-based trigger for redraws](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/4e43e36704f29acafa7ad4172e39594ce7976e38)")!

Reply:
Why not just have the top three checks and not do any of the existence checks? These existence checks will also fail in behavior when the ID = 0 since IDs are numbers.

Reply:
They wouldn't fire when the selection changed. This I believe was a discussion we had some time ago in that one thread with Matt, but I think svelte-wise this is necessary.